// globalpayments-api/lib/src/Services/CreditService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreditService = void 0;
const __1 = require("../");
class CreditService {
    constructor(config, configName = "default") {
        __1.ServicesContainer.configureService(config, configName);
    }
    authorize(amount) {
        return new __1.AuthorizationBuilder(__1.TransactionType.Auth).withAmount(amount);
    }
    capture(transactionId) {
        const ref = new __1.TransactionReference();
        ref.paymentMethodType = __1.PaymentMethodType.Credit;
        ref.transactionId = transactionId;
        return new __1.ManagementBuilder(__1.TransactionType.Capture).withPaymentMethod(ref);
    }
    charge(amount) {
        return new __1.AuthorizationBuilder(__1.TransactionType.Sale).withAmount(amount);
    }
    edit(transactionId) {
        const ref = new __1.TransactionReference();
        ref.paymentMethodType = __1.PaymentMethodType.Credit;
        if (transactionId) {
            ref.transactionId = transactionId;
        }
        return new __1.ManagementBuilder(__1.TransactionType.Edit).withPaymentMethod(ref);
    }
    refund(amount) {
        const ref = new __1.TransactionReference();
        ref.paymentMethodType = __1.PaymentMethodType.Credit;
        return new __1.AuthorizationBuilder(__1.TransactionType.Refund)
            .withAmount(amount)
            .withPaymentMethod(ref);
    }
    reverse(amount) {
        const ref = new __1.TransactionReference();
        ref.paymentMethodType = __1.PaymentMethodType.Credit;
        return new __1.AuthorizationBuilder(__1.TransactionType.Reversal)
            .withAmount(amount)
            .withPaymentMethod(ref);
    }
    verify() {
        return new __1.AuthorizationBuilder(__1.TransactionType.Verify);
    }
    void(transactionId) {
        const ref = new __1.TransactionReference();
        ref.paymentMethodType = __1.PaymentMethodType.Credit;
        ref.transactionId = transactionId;
        return new __1.ManagementBuilder(__1.TransactionType.Void).withPaymentMethod(ref);
    }
}
exports.CreditService = CreditService;
