// globalpayments-api/lib/src/Services/HostedService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HostedService = void 0;
const __1 = require("../");
class HostedService {
    config;
    constructor(config, configName = "default") {
        this.config = config;
        __1.ServicesContainer.configureService(config, configName);
    }
    authorize(amount) {
        return new __1.AuthorizationBuilder(__1.TransactionType.Auth).withAmount(amount);
    }
    charge(amount) {
        return new __1.AuthorizationBuilder(__1.TransactionType.Sale).withAmount(amount);
    }
    verify(amount) {
        return new __1.AuthorizationBuilder(__1.TransactionType.Verify).withAmount(amount);
    }
    parseResponse(json, encoded = true) {
        const response = JSON.parse(json);
        const decoder = encoded ? __1.StringUtils.atob : (t) => t;
        const timestamp = decoder(response.TIMESTAMP);
        const merchantId = decoder(response.MERCHANT_ID);
        const orderId = decoder(response.ORDER_ID);
        const result = decoder(response.RESULT);
        const message = decoder(response.MESSAGE);
        const transactionId = decoder(response.PASREF);
        const authCode = decoder(response.AUTHCODE);
        const sha1Hash = decoder(response.SHA1HASH);
        const hash = __1.GenerationUtils.generateHash([
            timestamp,
            merchantId,
            orderId,
            result,
            message,
            transactionId,
            authCode,
        ].join("."), this.config.sharedSecret);
        if (hash !== sha1Hash) {
            throw new __1.ApiError("Incorrect hash. Please check your code and the Developers Documentation.");
        }
        const transaction = new __1.Transaction();
        transaction.authorizedAmount = decoder(response.AMOUNT);
        transaction.cvnResponseCode = decoder(response.CVNRESULT);
        transaction.responseCode = result;
        transaction.responseMessage = message;
        transaction.avsResponseCode = decoder(response.AVSPOSTCODERESULT);
        transaction.transactionReference = new __1.TransactionReference();
        transaction.transactionReference.authCode = authCode;
        transaction.transactionReference.orderId = orderId;
        transaction.transactionReference.paymentMethodType =
            __1.PaymentMethodType.Credit;
        transaction.transactionReference.transactionId = transactionId;
        return transaction;
    }
}
exports.HostedService = HostedService;
