// globalpayments-api/lib/src/Services/ReportingService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReportingService = void 0;
/* eslint-disable indent */
const __1 = require("../");
class ReportingService {
    static activity() {
        return new __1.TransactionReportBuilder(__1.ReportType.Activity);
    }
    static transactionDetail(transactionId) {
        return new __1.TransactionReportBuilder(__1.ReportType.TransactionDetail).withTransactionId(transactionId);
    }
    static findTransactions() {
        return new __1.TransactionReportBuilder(__1.ReportType.FindTransactions);
    }
    static findTransactionsPaged(page, pageSize, transactionId) {
        return new __1.TransactionReportBuilder(__1.ReportType.FindTransactionsPaged)
            .withTransactionId(transactionId)
            .withPaging(page, pageSize);
    }
    static findStoredPaymentMethodsPaged(page, pageSize) {
        return new __1.TransactionReportBuilder(__1.ReportType.FindStoredPaymentMethodsPaged).withPaging(page, pageSize);
    }
    static storedPaymentMethodDetail(paymentMethodId) {
        return new __1.TransactionReportBuilder(__1.ReportType.StoredPaymentMethodDetail).withStoredPaymentMethodId(paymentMethodId);
    }
    static disputeDetail(disputeId) {
        return new __1.TransactionReportBuilder(__1.ReportType.DisputeDetail).withDisputeId(disputeId);
    }
    static findDisputesPaged(page, pageSize) {
        return new __1.TransactionReportBuilder(__1.ReportType.FindDisputesPaged).withPaging(page, pageSize);
    }
    static documentDisputeDetail(disputeId) {
        return new __1.TransactionReportBuilder(__1.ReportType.DocumentDisputeDetail).withDisputeId(disputeId);
    }
    static settlementDisputeDetail(settlementDisputeId) {
        return new __1.TransactionReportBuilder(__1.ReportType.SettlementDisputeDetail).withSettlementDisputeId(settlementDisputeId);
    }
    static findSettlementDisputesPaged(page, pageSize) {
        return new __1.TransactionReportBuilder(__1.ReportType.FindSettlementDisputesPaged).withPaging(page, pageSize);
    }
    static findSettlementTransactionsPaged(page, pageSize, transactionId) {
        return new __1.TransactionReportBuilder(__1.ReportType.FindSettlementTransactionsPaged)
            .withTransactionId(transactionId)
            .withPaging(page, pageSize);
    }
    static findDepositsPaged(page, pageSize) {
        return new __1.TransactionReportBuilder(__1.ReportType.FindDepositsPaged).withPaging(page, pageSize);
    }
    static depositDetail(depositId) {
        return new __1.TransactionReportBuilder(__1.ReportType.DepositDetail).withDepositId(depositId);
    }
}
exports.ReportingService = ReportingService;
