// globalpayments-api/lib/src/Terminals/UPA/Entities/UpaMessageId.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpaMessageId = void 0;
var UpaMessageId;
(function (UpaMessageId) {
    UpaMessageId["SALE"] = "Sale";
    UpaMessageId["VOID"] = "Void";
    UpaMessageId["REFUND"] = "Refund";
    UpaMessageId["EOD"] = "EODProcessing";
    UpaMessageId["SENDSAF"] = "SendSAF";
    UpaMessageId["TIPADJUST"] = "TipAdjust";
    UpaMessageId["CARD_VERIFY"] = "CardVerify";
    UpaMessageId["GET_SAF_REPORT"] = "GetSAFReport";
    UpaMessageId["CANCEL"] = "CancelTransaction";
    UpaMessageId["REBOOT"] = "Reboot";
    UpaMessageId["LINEITEM"] = "LineItemDisplay";
    UpaMessageId["REVERSAL"] = "Reversal";
    UpaMessageId["GET_BATCH_REPORT"] = "GetBatchReport";
    UpaMessageId["BALANCE_INQUIRY"] = "BalanceInquiry";
    UpaMessageId["PRE_AUTH"] = "PreAuth";
    UpaMessageId["AUTH_COMPLETION"] = "AuthCompletion";
    UpaMessageId["TOKENIZE"] = "Tokenize";
    UpaMessageId["GET_OPEN_TAB_DETAILS"] = "GetOpenTabDetails";
})(UpaMessageId || (exports.UpaMessageId = UpaMessageId = {}));
