// globalpayments-api/lib/src/Terminals/DeviceController.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeviceController = void 0;
class DeviceController {
    _interface;
    _settings;
    connector;
    connectionMode;
    deviceType;
    requestIdProvider;
    constructor(settings) {
        this._settings = settings;
        this.connector = this.configureConnector();
    }
    getConnectionMode() {
        if (this._settings != null) {
            return this._settings.connectionMode;
        }
        return null;
    }
    getDeviceType() {
        if (this._settings != null) {
            return this._settings.deviceType;
        }
        return null;
    }
    getRequestIdProvider() {
        if (this._settings != null) {
            return this._settings.requestIdProvider;
        }
        return null;
    }
    send(message, requestType) {
        message.awaitResponse = true;
        if (this.connector) {
            return this.connector.send(message);
        }
        requestType;
    }
}
exports.DeviceController = DeviceController;
