// globalpayments-api/lib/src/Terminals/DeviceMessage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeviceMessage = void 0;
class DeviceMessage {
    keepAlive;
    awaitResponse;
    buffer;
    jsonRequest;
    constructor(buffer) {
        this.buffer = buffer;
    }
    getSendBuffer() {
        return this.buffer;
    }
    setJsonRequest(jsonRequest) {
        this.jsonRequest = jsonRequest;
    }
    getJsonRequest() {
        return this.jsonRequest;
    }
    toString() {
        return this.buffer.toString();
    }
    getRequestField(key) {
        return this.iterateDeep(this.jsonRequest, key);
    }
    iterateDeep(obj, key) {
        if (!this.jsonRequest || typeof this.jsonRequest !== "object") {
            return undefined;
        }
        if (obj[key]) {
            return obj[key];
        }
        for (const objectKey in obj) {
            if (typeof obj[objectKey] === "object" && obj[objectKey] !== null) {
                return this.iterateDeep(obj[objectKey], key);
            }
        }
    }
}
exports.DeviceMessage = DeviceMessage;
