// globalpayments-api/lib/src/Terminals/SummaryResponse.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SummaryResponse = exports.SummaryType = void 0;
var SummaryType;
(function (SummaryType) {
    SummaryType[SummaryType["Approved"] = 0] = "Approved";
    SummaryType[SummaryType["PartiallyApproved"] = 1] = "PartiallyApproved";
    SummaryType[SummaryType["VoidApproved"] = 2] = "VoidApproved";
    SummaryType[SummaryType["Pending"] = 3] = "Pending";
    SummaryType[SummaryType["VoidPending"] = 4] = "VoidPending";
    SummaryType[SummaryType["Declined"] = 5] = "Declined";
    SummaryType[SummaryType["VoidDeclined"] = 6] = "VoidDeclined";
    SummaryType[SummaryType["OfflineApproved"] = 7] = "OfflineApproved";
    SummaryType[SummaryType["Provsional"] = 8] = "Provsional";
    SummaryType[SummaryType["Discarded"] = 9] = "Discarded";
    SummaryType[SummaryType["VoidProvisional"] = 10] = "VoidProvisional";
    SummaryType[SummaryType["VoidDiscarded"] = 11] = "VoidDiscarded";
    SummaryType[SummaryType["Reversal"] = 12] = "Reversal";
    SummaryType[SummaryType["EmvDeclined"] = 13] = "EmvDeclined";
    SummaryType[SummaryType["Attachment"] = 14] = "Attachment";
    SummaryType[SummaryType["Unknown"] = 15] = "Unknown";
})(SummaryType || (exports.SummaryType = SummaryType = {}));
class SummaryResponse {
    amount;
    amountDue;
    authorizedAmount;
    count;
    summaryType;
    totalAmountCount;
    transactions;
}
exports.SummaryResponse = SummaryResponse;
