// globalpayments-api/lib/src/Terminals/UPA/Reponses/TransactionResponse.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionResponse = void 0;
const src_1 = require("../../../../src");
class TransactionResponse {
    availableBalance;
    transactionId;
    terminalRefNumber;
    token;
    cardBrandTransId;
    signatureStatus;
    signatureData;
    transactionType;
    maskedCardNumber;
    entryMethod;
    authorizationCode;
    transactionAmount;
    amountDue;
    balanceAmount;
    cardBIN;
    cardPresent;
    expirationDate;
    avsResponseCode;
    avsResponseText;
    cvvResponseCode;
    cvvResponseText;
    taxExempt;
    taxExemptId;
    ticketNumber;
    paymentType;
    applicationPreferredName;
    applicationLabel;
    applicationId;
    applicationCryptogramType;
    applicationCryptogram;
    cardHolderVerificationMethod;
    terminalVerificationResults;
    merchantFee;
    status;
    command;
    version;
    deviceResponseCode;
    deviceResponseText;
    responseCode;
    responseText;
    approvalCode;
    tipAmount;
    baseAmount;
    cashBackAmount;
    referenceNumber;
    cardHolderName;
    requestId;
    constructor(jsonResponse) {
        if (this.isGpApiResponse(jsonResponse)) {
            this.requestId = this.transactionId = jsonResponse.id;
            this.deviceResponseText = jsonResponse.status;
            this.responseText = this.deviceResponseCode =
                jsonResponse.action.result_code;
            this.deviceResponseText = jsonResponse.status;
        }
        else {
            throw new src_1.NotImplementedError();
        }
    }
    isGpApiResponse(jsonResponse) {
        return !!(jsonResponse.provider && jsonResponse.provider === src_1.GatewayProvider.GpApi);
    }
}
exports.TransactionResponse = TransactionResponse;
