// globalpayments-api/lib/src/PaymentMethods/CreditCardData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreditCardData = void 0;
const __1 = require("../");
const Credit_1 = require("./Credit");
class CreditCardData extends Credit_1.Credit {
    static cardTypes = [
        { name: "Visa", regex: /^4/ },
        { name: "MC", regex: /^(5[1-5]|2[2-7])/ },
        { name: "Amex", regex: /^3[47]/ },
        { name: "Diners", regex: /^3[0689]/ },
        { name: "EnRoute", regex: /^2(014|149)/ },
        { name: "Discover", regex: /^6([045]|22)/ },
        { name: "Jcb", regex: /^35/ },
    ];
    number;
    expMonth;
    expYear;
    cvn;
    cvnPresenceIndicator;
    cardHolderName;
    cardPresent;
    readerPresent;
    constructor() {
        super();
        this.cardPresent = false;
        this.readerPresent = false;
        this.cvnPresenceIndicator = __1.CvnPresenceIndicator.NotRequested;
    }
    getCardType() {
        const number = this.number.replace(" ", "").replace("-", "");
        for (const type of CreditCardData.cardTypes) {
            if (type.regex.test(number)) {
                return type.name;
            }
        }
        return "Unknown";
    }
    getShortExpiry() {
        if (this.expMonth != null && this.expYear != null) {
            return this.expMonth + this.expYear.slice(-2);
        }
        return "";
    }
}
exports.CreditCardData = CreditCardData;
