// globalpayments-api/lib/src/PaymentMethods/CreditTrackData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreditTrackData = void 0;
const __1 = require("../");
const Credit_1 = require("./Credit");
class CreditTrackData extends Credit_1.Credit {
    entryMethod;
    value;
    discretionaryData;
    expiry;
    pan;
    purchaseDeviceSequenceNumber;
    trackNumber;
    trackData;
    setTrackData(value) {
        if (!this.value) {
            this.setValue(value);
        }
        else {
            this.trackData = value;
        }
    }
    setValue(value) {
        this.value = value;
        __1.CardUtils.parseTrackData(this);
        this.cardType = __1.CardUtils.getCardType(this.pan);
        this.isFleet = __1.CardUtils.isFleet(this.cardType, this.pan);
        if (this.cardType === "WexFleet" &&
            this.discretionaryData !== null &&
            this.discretionaryData.length >= 8) {
            this.purchaseDeviceSequenceNumber = this.discretionaryData.substring(3, 11);
        }
    }
}
exports.CreditTrackData = CreditTrackData;
