// globalpayments-api/lib/src/PaymentMethods/Debit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Debit = void 0;
const __1 = require("../");
const PaymentMethod_1 = require("./PaymentMethod");
class Debit extends PaymentMethod_1.PaymentMethod {
    encryptionData;
    paymentMethodType = __1.PaymentMethodType.Debit;
    pinBlock;
    /**
     * Authorizes the payment method and captures the entire authorized amount
     *
     * @param string|number amount Amount to authorize
     *
     * @return AuthorizationBuilder
     */
    charge(amount) {
        return new __1.AuthorizationBuilder(__1.TransactionType.Sale, this).withAmount(amount);
    }
    /**
     * Adds value to the payment method
     *
     * @param string|number amount Amount to add
     *
     * @return AuthorizationBuilder
     */
    addValue(amount) {
        return new __1.AuthorizationBuilder(__1.TransactionType.AddValue, this).withAmount(amount);
    }
    /**
     * Refunds the payment method
     *
     * @param string|number amount Amount to refund
     *
     * @return AuthorizationBuilder
     */
    refund(amount) {
        return new __1.AuthorizationBuilder(__1.TransactionType.Refund, this).withAmount(amount);
    }
    /**
     * Reverses the payment method
     *
     * @param string|number amount Amount to reverse
     *
     * @return AuthorizationBuilder
     */
    reverse(amount) {
        return new __1.AuthorizationBuilder(__1.TransactionType.Reversal, this).withAmount(amount);
    }
    /**
     * Authorizes the payment method
     *
     * @param string|number amount Amount to reverse
     *
     * @return AuthorizationBuilder
     */
    authorize(amount) {
        return new __1.AuthorizationBuilder(__1.TransactionType.Auth, this)
            .withAmount(amount)
            .withAmountEstimated(true);
    }
}
exports.Debit = Debit;
