// globalpayments-api/lib/src/PaymentMethods/PaymentMethod.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaymentMethod = void 0;
const __1 = require("../");
class PaymentMethod {
    paymentMethodType;
    get isAuthable() {
        return this.authorize !== undefined;
    }
    get isBalanceable() {
        return this.balanceInquiry !== undefined;
    }
    get isCardData() {
        return (((this.isTokenizable &&
            this.token !== undefined) ||
            this.number !== undefined) &&
            this.paymentMethodType !== __1.PaymentMethodType.Gift);
    }
    get isChargable() {
        return this.charge !== undefined;
    }
    get isEditable() {
        return this.edit !== undefined;
    }
    get isEncryptable() {
        return this.encryptionData !== undefined;
    }
    get isPinProtected() {
        return this.pinBlock !== undefined;
    }
    get isPrePayable() {
        return this.addValue !== undefined;
    }
    get isRefundable() {
        return this.refund !== undefined;
    }
    get isReversable() {
        return this.reverse !== undefined;
    }
    get isTokenizable() {
        return this.tokenize !== undefined;
    }
    get isTrackData() {
        return (this.value !== undefined &&
            (this.entryMethod !== undefined ||
                this.pinBlock !== undefined ||
                this.encryptionData !== undefined));
    }
    get isVerifyable() {
        return this.verify !== undefined;
    }
    get isVoidable() {
        return this.void !== undefined;
    }
    get isSecure3d() {
        return this !== undefined;
    }
    get isSecureCheck() {
        return this !== undefined;
    }
}
exports.PaymentMethod = PaymentMethod;
