// globalpayments-api/lib/src/Builders/TransactionBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionBuilder = void 0;
const __1 = require("../");
const BaseBuilder_1 = require("./BaseBuilder");
class TransactionBuilder extends BaseBuilder_1.BaseBuilder {
    paymentMethod;
    transactionType;
    transactionModifier = __1.TransactionModifier.None;
    supplementaryData = {};
    constructor(type, paymentMethod) {
        super();
        this.transactionType = type;
        if (paymentMethod) {
            this.paymentMethod = paymentMethod;
        }
    }
    withModifier(modifier) {
        if (modifier !== undefined) {
            this.transactionModifier = modifier;
        }
        return this;
    }
    withSupplementaryData(key, value) {
        if (key instanceof Object) {
            for (const supplementaryDataKey of Object.keys(key)) {
                this.withSupplementaryData(supplementaryDataKey, key[supplementaryDataKey]);
            }
        }
        if (key && typeof key === "string" && !!value) {
            this.supplementaryData[key] = value;
        }
        return this;
    }
    withAllowDuplicates(allowDuplicates) {
        this.allowDupplicates = allowDuplicates;
        return this;
    }
}
exports.TransactionBuilder = TransactionBuilder;
