// globalpayments-api/lib/src/ServiceConfigs/AcceptorConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AcceptorConfig = void 0;
const Entities_1 = require("../../src/Entities");
class AcceptorConfig {
    /**
     * Used w/TransIT
     */
    cardCaptureCapability;
    /**
     * Used w/TransIT
     */
    cardDataInputCapability;
    /**
     * Used w/TransIT
     */
    cardDataOutputCapability;
    /**
     * Used w/TransIT; corresponding tag will default to eComm or Manual if this isn't used
     */
    cardDataSource;
    /**
     * Used w/TransIT
     */
    cardHolderAuthenticationCapability;
    /**
     * Used w/TransIT
     */
    cardHolderAuthenticationEntity;
    constructor(cardCaptureCapability = false, cardDataInputCapability = Entities_1.CardDataInputCapability.KeyedEntryOnly, cardDataOutputCapability = Entities_1.CardDataOutputCapability.None, cardHolderAuthenticationCapability = Entities_1.CardHolderAuthenticationCapability.NoCapability, cardHolderAuthenticationEntity = Entities_1.CardHolderAuthenticationEntity.NotAuthenticated) {
        this.cardCaptureCapability = cardCaptureCapability;
        this.cardDataInputCapability = cardDataInputCapability;
        this.cardDataOutputCapability = cardDataOutputCapability;
        this.cardHolderAuthenticationCapability =
            cardHolderAuthenticationCapability;
        this.cardHolderAuthenticationEntity = cardHolderAuthenticationEntity;
    }
    validate() {
        // for use in future gateway integrations
    }
}
exports.AcceptorConfig = AcceptorConfig;
