// globalpayments-api/lib/src/Entities/ServiceEndpoints.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServiceEndpoints = void 0;
var ServiceEndpoints;
(function (ServiceEndpoints) {
    ServiceEndpoints["GLOBAL_ECOM_PRODUCTION"] = "https://api.realexpayments.com/epage-remote.cgi";
    ServiceEndpoints["GLOBAL_ECOM_TEST"] = "https://api.sandbox.realexpayments.com/epage-remote.cgi";
    ServiceEndpoints["GP_API_PRODUCTION"] = "https://apis.globalpay.com/ucp";
    ServiceEndpoints["GP_API_TEST"] = "https://apis.sandbox.globalpay.com/ucp";
    ServiceEndpoints["PORTICO_PRODUCTION"] = "https://api2.heartlandportico.com";
    ServiceEndpoints["PORTICO_TEST"] = "https://cert.api2.heartlandportico.com";
    ServiceEndpoints["PROPAY_TEST"] = "https://xmltest.propay.com/API/PropayAPI.aspx";
    ServiceEndpoints["PROPAY_TEST_CANADIAN"] = "https://xmltestcanada.propay.com/API/PropayAPI.aspx";
    ServiceEndpoints["PROPAY_PRODUCTION"] = "https://epay.propay.com/API/PropayAPI.aspx";
    ServiceEndpoints["PROPAY_PRODUCTION_CANADIAN"] = "https://www.propaycanada.ca/API/PropayAPI.aspx";
})(ServiceEndpoints || (exports.ServiceEndpoints = ServiceEndpoints = {}));
