// globalpayments-api/lib/src/Builders/BaseBuilder/ValidationTarget.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidationTarget = void 0;
const ValidationClause_1 = require("./ValidationClause");
class ValidationTarget {
    parent;
    type;
    property;
    clause;
    constraint;
    constraintProperty;
    enumName;
    precondition;
    constructor(parent, enumName, type) {
        this.parent = parent;
        this.type = type;
        this.enumName = enumName;
    }
    with(property, constraint) {
        this.constraintProperty = property;
        this.constraint = constraint;
        return this;
    }
    check(targetProperty) {
        this.property = targetProperty;
        this.clause = new ValidationClause_1.ValidationClause(this.parent, this);
        return this.clause;
    }
    when(targetProperty) {
        this.property = targetProperty;
        this.precondition = new ValidationClause_1.ValidationClause(this.parent, this, true);
        return this.precondition;
    }
}
exports.ValidationTarget = ValidationTarget;
