// globalpayments-api/lib/src/ConfiguredServices.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfiguredServices = void 0;
const Entities_1 = require("./Entities");
class ConfiguredServices {
    payFacProvider;
    gatewayConnector;
    recurringConnector;
    reportingService;
    deviceInterface;
    _deviceController;
    secure3dProviders = new Map();
    constructor() { }
    /**
     * @return void
     */
    setPayFacProvider(provider) {
        this.payFacProvider = provider;
    }
    /**
     * @return IPayFacProvider
     */
    getPayFacProvider() {
        return this.payFacProvider || null;
    }
    get deviceController() {
        return this._deviceController;
    }
    set deviceController(deviceController) {
        this._deviceController = deviceController;
        this.deviceInterface = deviceController.configureInterface();
    }
    getSecure3dProvider(version) {
        if (this.secure3dProviders.get(version)) {
            return this.secure3dProviders.get(version);
        }
        else if (version === Entities_1.Secure3dVersion.ANY) {
            let provider = this.secure3dProviders.get(Entities_1.Secure3dVersion.TWO);
            if (provider === null) {
                provider = this.secure3dProviders.get(Entities_1.Secure3dVersion.ONE);
            }
            return provider;
        }
        else {
            return null;
        }
    }
    setSecure3dProvider(version, provider) {
        this.secure3dProviders.set(version, provider);
    }
}
exports.ConfiguredServices = ConfiguredServices;
